autonomous/auto_base.py

```diff
@@ -0,0 +1,157 @@
+import math
+
+import choreo
+from choreo.trajectory import SwerveSample
+from magicbot import AutonomousStateMachine, state
+from wpilib import RobotBase
+from wpimath.controller import PIDController
+from wpimath.geometry import Pose2d
+from wpimath.kinematics import ChassisSpeeds
+
+from components.algae_manipulator import AlgaeManipulatorComponent
+from components.chassis import ChassisComponent
+from controllers.algae_shooter import AlgaeShooter
+from controllers.coral_placer import CoralPlacer
+from controllers.reef_intake import ReefIntake
+from utilities import game
+
+
+class AutoBase(AutonomousStateMachine):
+    algae_shooter: AlgaeShooter
+    coral_placer: CoralPlacer
+    reef_intake: ReefIntake
+
+    algae_manipulator_component: AlgaeManipulatorComponent
+    chassis: ChassisComponent
+
+    DISTANCE_TOLERANCE = 0.05  # metres
+
+    def __init__(self, trajectory_names: list[str]) -> None:
+        # We want to parameterise these by paths and potentially a sequence of events
+        super().__init__()
+        self.x_controller = PIDController(1.0, 0.0, 0.0)
+        self.y_controller = PIDController(1.0, 0.0, 0.0)
+        self.heading_controller = PIDController(1.0, 0, 0)
+        self.heading_controller.enableContinuousInput(-math.pi, math.pi)
+
+        self.current_leg = -1
+        self.starting_pose = None
+        self.trajectories = []
+        for trajectory_name in trajectory_names:
+            try:
+                self.trajectories.append(choreo.load_swerve_trajectory(trajectory_name))
+                if self.starting_pose is None:
+                    self.starting_pose = self.get_starting_pose()
```

Comment:
I thought I deleted this in #211?

Reply:
I think you might have and then rebase shenanigans ensued. I am pretty sure that setting the `starting_pose` class member is required now because of how the LEDs interact with auto though.